include/mbedtls/mps/writer.h

```diff
@@ -0,0 +1,694 @@
+ *
+ * \return          \c 0 on success.
+ * \return          A negative error code \c MBEDTLS_ERR_WRITER_XXX on failure.
+ *
+ */
+/*@
+  WRITER_EXT_INV_REQUIRES( writer )
+  @*/
+int mbedtls_writer_free_ext( mbedtls_writer_ext *writer );
+
+/**
+ * \brief           Request buffer to hold outbound data.
+ *
+ * \param writer    The extended writer context to use.
+ * \param desired   The desired size of the outgoing data buffer.
+ * \param buffer    The address at which to store the address
+ *                  of the outgoing data buffer on success.
+ * \param buflen    The address at which to store the actual
+ *                  size of the outgoing data buffer on success.
+ *                  May be NULL (see below).
+ *
+ * \note            If \p buflen is \c NULL, the function fails
+ *                  if it cannot provide an outgoing data buffer
+ *                  of the requested size \p desired.
+ *
+ * \return          \c 0 on success. In this case, \c *buf holds the
+ *                  address of a buffer of size \c *buflen
+ *                  (if \c buflen != NULL) or \p desired
+ *                  (if \c buflen is \c NULL).
+ * \return          #MBEDTLS_ERR_WRITER_BOUNDS_VIOLATION if the write
+ *                  request exceeds the bounds of the current group.
+ *
+ */
+
+/*@
+  requires \valid( buffer );
+  requires ( buflen == NULL ) || \valid( buflen );
+  WRITER_EXT_INV_REQUIRES(writer)
+  WRITER_EXT_INV_ENSURES(writer)
+  @*/
+int mbedtls_writer_get_ext( mbedtls_writer_ext *writer, size_t desired,
+                            unsigned char **buffer, size_t *buflen );
+
+/**
+ * \brief           Signal that all output buffers previously obtained
+ *                  from mbedtls_writer_get() are ready to be dispatched.
+ *
+ * \param writer    The extended writer context to use.
+ *
+ * \note            After this function has been called, all
+ *                  output buffers obtained from prior calls to
+ *                  mbedtls_writer_get() are invalid and must not
+ *                  be accessed anymore.
+ *
+ * \return          \c 0 on success.
+ * \return          A negative error code \c MBEDTLS_ERR_WRITER_XXX on failure.
+ *
+ */
+/*@
+  WRITER_EXT_INV_REQUIRES(writer)
+  WRITER_EXT_INV_ENSURES(writer)
+  @*/
+int mbedtls_writer_commit_ext( mbedtls_writer_ext *writer );
+
+/**
+ * \brief            Open a new logical subbuffer.
+ *
+ * \param writer     The extended writer context to use.
+ * \param group_size The offset of the end of the subbuffer
+ *                   from the end of the last successful fetch.
+ *
+ * \return           \c 0 on success.
+ * \return           #MBEDTLS_ERR_WRITER_BOUNDS_VIOLATION if
+ *                   the new group is not contained in the
+ *                   current group. In this case, the extended
+ *                   writer is unchanged and hence remains intact.
+ * \return           Another negative error code otherwise.
```

Comment:
@AndrzejKurek Could you elaborate?

Reply:
Unfortunately I can't check the version that I've commented on (force push?), but on the latest version it looks consistent with surrounding functions, so we can dismiss my comment on that.

Reply:
I don't remember if there has been a force push initially after you had already started reviewing - if so, sorry for that! As long as reviewing continues, I try to use incremental commits, leaving a potential history cleanup to the end. That said, sometimes a force push might be necessary because of the number and relation between the various MPS branches (HS rework, Record Enc/Dec, Reader/Writer, L1-L4).